002-tidyr-stringr.Rmd

```diff
@@ -0,0 +1,256 @@
+
+Should these functions use stringr or not?
+It's nice to keep dependencies low, but this will considerably increase the cost of implementation because they'll need both stringr and base R implementations.
+
+Before committing to implementation a pure base R solution, I think we should discover if this is a real need.
+To do this, I suggest we take stringr as a suggested dependency that's required by `separate_*()`.
+We'll open an issue to track how much pain this causes for people, and consider a base R implementation in a future release if there's significant need.
+
+## Alternatives
+
+These alternatives were considered and rejected.
+
+### One function per column
+
+Another option is to divide the problem into two parts:
+
+| Into            | stringr                 | tidyr               |
+|-----------------|-------------------------|---------------------|
+| Named/fixed     | `str_separate_wider()`  | `separate_wider()`  |
+| Unnamed/varying | `str_separate_longer()` | `separate_longer()` |
+
+Then each function has a `split` argument which can be one of the following cases:
+
+-   `by_position(1, 2, 3)`
+-   `by_width(1, 2, 3)`
+-   `by_separator(",", ignore.case = TRUE)`
+-   `by_regex_group("a(.)c")` (`by_group()` is too short/ambiguous).
+
+(This could also be defined by 3-4 mutual exclusive arguments, but that is hard to extend and hard to document)
+
+Pros:
+
+-   8 functions instead of 10, and extending with more splitting types is easy.
+-   Composing functions in this way seems more "tidyverse" like.
+
+Cons:
+
+-   Composing the functions on your own makes it harder to see the full set of options, and there's not enough total options to make the flexibility really worth it.
+-   Will be tempting to use `someotherfunction(x, by_position(1, 2, 3))`.
+
+Overall, this proposal only reduces the total number of functions by two, in return for a small gain.
+
+### Eliminate tidyr functions
+
+Can we eliminate the tidyr wrappers altogether and just use dplyr + stringr?
+This seemed like the "obvious" solution to me at least twice, so I wanted to make sure to explain why it doesn't work.
+
+To make the discussion concrete I'll start with quick and dirty implementations of `str_separate_by_wider()` and `str_separate_by_longer()`:
+
+```{r}
+str_separate_by_wider <- function(x, pattern, into) {
+  mat <- str_split_fixed(x, pattern, length(into))
+  colnames(mat) <- into
+  tibble::as_tibble(mat)
+}
+str_separate_by_longer <- function(x, pattern) {
+  str_split(x, pattern)
+}
+
+str_separate_by_wider(c("a-b", "c-d"), "-", c("first", "second"))
+str_separate_by_longer(c("a-b", "c-d-e"), "-")
+```
+
+We can use `str_separate_by_wider()` directly with `mutate()` to achieve the same affect as `separate()`:
+
+```{r}
+library(dplyr, warn.conflicts = FALSE)
+
+df <- tibble(x = c("a-b", "c-d"))
+df %>% separate(x, c("first", "second"), "-")
+
+df %>% 
+  mutate(
+    str_separate_by_wider(x, "-", c("first", "second")), 
+    .keep = "unused"
+  )
```

Comment:
And here is my multi-column approach with `str_separate_by_wider()`

``` r
library(tidyverse)
library(stringr)

df <- tibble(
  x = 1:2,
  y = c("a-c", "c-d"),
  z = c("q-r", "z-f")
)

str_separate_by_wider <- function(x, pattern, into) {
  mat <- str_split_fixed(x, pattern, length(into))
  colnames(mat) <- into
  tibble::as_tibble(mat)
}

df %>%
  mutate(across(c(y, z), ~str_separate_by_wider(.x, "-", c("first", "second")))) %>%
  unpack(c(y, z), names_sep = "_")
#> # A tibble: 2 × 5
#>       x y_first y_second z_first z_second
#>   <int> <chr>   <chr>    <chr>   <chr>   
#> 1     1 a       c        q       r       
#> 2     2 c       d        z       f
```

<sup>Created on 2021-12-15 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>